JDK-8306510: Print number of threads and stack sizes in error reports

tstuefe opened the pull request:
To analyze OOM situations, number of threads and their stack sizes are useful. NMT tells us, but NMT may be disabled.

We already print threads and stacks. Today one could count lines and calculate stack sizes via stack base and end address.

But to save time, it would be helpful to see the number of threads and their stack sizes printed out. 

---------

Patch does that, and also adjusts indentation a bit to make the output more readable. I'm open to leaving out that part if people object. The patch only affects printing hs-err files, no other printouts should be affected (I tested jcmd Thread.print and output is unchanged).

gerard-ziemski reviewed (changes requested):
Looks like a nice addition.

I have 2 pieces of feedback:

#1 In your example you have

```
Other Threads:
  0x00007f0ac4290f00 VMThread "VM Thread"                    [stack: 0x00007f0a98141000,0x00007f0a98241000] (1024K) [id=239095]
  0x00007f0ac443c8a0 WatcherThread "VM Periodic Task Thread" [stack: 0x00007f0a88eb6000,0x00007f0a88fb6000] (1024K) [id=239107]
  0x00007f0ac405aeb0 WorkerThread "GC Thread#0"              [stack: 0x00007f0ac8929000,0x00007f0ac8a29000] (1024K) [id=239090]
  0x00007f0ac406c320 ConcurrentGCThread "G1 Main Marker"     [stack: 0x00007f0ac8827000,0x00007f0ac8927000] (1024K) [id=239091]
  0x00007f0ac406d490 WorkerThread "G1 Conc#0"                [stack: 0x00007f0ac8725000,0x00007f0ac8825000] (1024K) [id=239092]
  0x00007f0ac412eae0 ConcurrentGCThread "G1 Refine#0"        [stack: 0x00007f0a98462000,0x00007f0a98562000] (1024K) [id=239093]
  0x00007f0ac412fc40 ConcurrentGCThread "G1 Service"         [stack: 0x00007f0a98360000,0x00007f0a98460000] (1024K) [id=239094]
Total: 8
```
Claims that there are 8 total threads, but only 7 shown here?

#2 There is a difference on how you report thread details between `Java Threads` and `Other Threads`

```
Java Threads:
0x00007f0ac4028d50 JavaThread "main"                              [_thread_in_vm, id=239089, stack(0x00007f0aca01d000,0x00007f0aca11e000) (1028K)]

Other Threads:
  0x00007f0ac4290f00 VMThread "VM Thread"                    [stack: 0x00007f0a98141000,0x00007f0a98241000] (1024K) [id=239095]

```
Why can't we have more consistent formatting, something like:

```
Java Threads:
0x00007f0ac4028d50 JavaThread "main"                              [_thread_in_vm, id=239089, stack(0x00007f0aca01d000,0x00007f0aca11e000) (1028K)]

Other Threads:
  0x00007f0ac4290f00 VMThread "VM Thread"                    [id=239095, stack(0x00007f0a98141000,0x00007f0a98241000) (1024K)]
```

tstuefe commented:
Thank you Gerard.

> Looks like a nice addition.
> 
> I have 2 pieces of feedback:
> 
> 
> Claims that there are 8 total threads, but only 7 shown here?

Good catch. Fixed.

> 
> #2 There is a difference on how you report thread details between `Java Threads` and `Other Threads`

Yes, I originally wanted to preserve the format as much as possible for the sake of parsers, but I already added indentation... may as well hang for a sheep. Fixed.

dholmes-ora commented on src/hotspot/share/runtime/javaThread.cpp:
```diff
@@ -1515,21 +1515,23 @@ void JavaThread::print_on_error(outputStream* st, char *buf, int buflen) const {
-            p2i(stack_end()), p2i(stack_base()));
+  st->print(", stack(" PTR_FORMAT "," PTR_FORMAT ") (" PROPERFMT ")",
+            p2i(stack_end()), p2i(stack_base()),
+            PROPERFMTARGS((size_t)(stack_base() - stack_end())));
```
Please use `stack_size()`

dholmes-ora commented on src/hotspot/share/runtime/thread.cpp:
```diff
@@ -483,10 +483,11 @@ void Thread::print_on_error(outputStream* st, char* buf, int buflen) const {
-      st->print(" [id=%d]", osthread()->thread_id());
+      st->print(" [id=%d, stack:(" PTR_FORMAT "," PTR_FORMAT ") (" PROPERFMT ")]",
+                osthread()->thread_id(), p2i(stack_end()), p2i(stack_base()),
+                PROPERFMTARGS((size_t)(stack_base() - stack_end())));
```
Please use `stack_size()`

dholmes-ora commented on src/hotspot/share/runtime/threads.cpp:
```diff
@@ -1317,39 +1322,49 @@ void Threads::print_on_error(outputStream* st, Thread* current, char* buf,
   st->cr();
 
   st->print_cr("Other Threads:");
+  unsigned num_other = (VMThread::vm_thread() ? 1 : 0) + (WatcherThread::watcher_thread() ? 1 : 0) + (AsyncLogWriter::instance() ? 1 : 0);
```
Nit: no implicit booleans - these should be comparisons against `nullptr`.

dholmes-ora reviewed (changes requested):
Seems okay in principle, some changes requested.

Thanks.

dholmes-ora reviewed (approved):
Seems fine. Thanks.

tstuefe commented:
@dholmes-ora Thanks!
@gerard-ziemski did I address your feedback?

gerard-ziemski commented:
I still see a difference in how we report the thread info:

```
Java Threads: ( => current thread ):
  0x00007f0ac443a4b0 JavaThread "Notification Thread"        daemon [_thread_blocked, id=239106, stack(0x00007f0a88fb6000,0x00007f0a890b7000) (1028K)]
```

vs

```
Other Threads:
  0x00007f0ac4290f00 VMThread "VM Thread"                    [stack: 0x00007f0a98141000,0x00007f0a98241000] (1024K) [id=239095]
```

Did you decide not to pursue it?

Whoever wants to parse the output would have an easier time to re-use pattern for matching, if we made the threads info same.

tstuefe commented:
> I still see a difference in how we report the thread info:
> 
> ```
> Java Threads: ( => current thread ):
>   0x00007f0ac443a4b0 JavaThread "Notification Thread"        daemon [_thread_blocked, id=239106, stack(0x00007f0a88fb6000,0x00007f0a890b7000) (1028K)]
> ```
> 
> vs
> 
> ```
> Other Threads:
>   0x00007f0ac4290f00 VMThread "VM Thread"                    [stack: 0x00007f0a98141000,0x00007f0a98241000] (1024K) [id=239095]
> ```
> 
> Did you decide not to pursue it?
> 

I did though

https://github.com/openjdk/jdk/pull/13556/commits/b109cba11ed420a7f1bc4c05d25a43f6905b8113

```
163 Java Threads: ( => current thread )
164 =>0x00007f9328028d10 JavaThread "main"                              [_thread_in_vm, id=104806, stack(0x00007f932fdf0000,0x00007f932fef1000) (1028K)]
165   0x00007f932826dce0 JavaThread "Reference Handler"          daemon [_thread_blocked, id=104813, stack(0x00007f92e780d000,0x00007f92e790e000) (1028K)]
166   0x00007f932826f4c0 JavaThread "Finalizer"                  daemon [_thread_blocked, id=104814, stack(0x00007f92e770c000,0x00007f92e780d000) (1028K)]
167   0x00007f9328271300 JavaThread "Signal Dispatcher"          daemon [_thread_blocked, id=104815, stack(0x00007f92e760b000,0x00007f92e770c000) (1028K)]
168   0x00007f9328272b20 JavaThread "Service Thread"             daemon [_thread_blocked, id=104816, stack(0x00007f92e750a000,0x00007f92e760b000) (1028K)]
169   0x00007f9328274200 JavaThread "Monitor Deflation Thread"   daemon [_thread_blocked, id=104817, stack(0x00007f92e7409000,0x00007f92e750a000) (1028K)]
170   0x00007f9328275ef0 JavaThread "C2 CompilerThread0"         daemon [_thread_blocked, id=104818, stack(0x00007f92e7308000,0x00007f92e7409000) (1028K)]
171   0x00007f9328277780 JavaThread "C1 CompilerThread0"         daemon [_thread_blocked, id=104819, stack(0x00007f92e7207000,0x00007f92e7308000) (1028K)]
172   0x00007f932829f370 JavaThread "Notification Thread"        daemon [_thread_blocked, id=104820, stack(0x00007f92e7036000,0x00007f92e7137000) (1028K)]
173 Total: 9
174 
175 Other Threads:
176   0x00007f932822a3e0 VMThread "VM Thread"                           [id=104812, stack:(0x00007f92e7e7f000,0x00007f92e7f7f000) (1024K)]
177   0x00007f93282a1380 WatcherThread "VM Periodic Task Thread"        [id=104821, stack:(0x00007f92e2f00000,0x00007f92e3000000) (1024K)]
178   0x00007f932805add0 WorkerThread "GC Thread#0"                     [id=104807, stack:(0x00007f932c79c000,0x00007f932c89c000) (1024K)]
179   0x00007f932806c240 ConcurrentGCThread "G1 Main Marker"            [id=104808, stack:(0x00007f932c69b000,0x00007f932c79b000) (1024K)]
180   0x00007f932806d3b0 WorkerThread "G1 Conc#0"                       [id=104809, stack:(0x00007f932c59a000,0x00007f932c69a000) (1024K)]
181   0x00007f932812ea00 ConcurrentGCThread "G1 Refine#0"               [id=104810, stack:(0x00007f92fc1a6000,0x00007f92fc2a6000) (1024K)]
182   0x00007f932812fb60 ConcurrentGCThread "G1 Service"                [id=104811, stack:(0x00007f92fc0a5000,0x00007f92fc1a5000) (1024K)]
183 Total: 7
```

gerard-ziemski commented:
> > I still see a difference in how we report the thread info:
> > ```
> > Java Threads: ( => current thread ):
> >   0x00007f0ac443a4b0 JavaThread "Notification Thread"        daemon [_thread_blocked, id=239106, stack(0x00007f0a88fb6000,0x00007f0a890b7000) (1028K)]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > vs
> > ```
> > Other Threads:
> >   0x00007f0ac4290f00 VMThread "VM Thread"                    [stack: 0x00007f0a98141000,0x00007f0a98241000] (1024K) [id=239095]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Did you decide not to pursue it?
> 
> I did though
> 
> [b109cba](https://github.com/openjdk/jdk/commit/b109cba11ed420a7f1bc4c05d25a43f6905b8113)
> 
> ```
> 163 Java Threads: ( => current thread )
> 164 =>0x00007f9328028d10 JavaThread "main"                              [_thread_in_vm, id=104806, stack(0x00007f932fdf0000,0x00007f932fef1000) (1028K)]
> 165   0x00007f932826dce0 JavaThread "Reference Handler"          daemon [_thread_blocked, id=104813, stack(0x00007f92e780d000,0x00007f92e790e000) (1028K)]
> 166   0x00007f932826f4c0 JavaThread "Finalizer"                  daemon [_thread_blocked, id=104814, stack(0x00007f92e770c000,0x00007f92e780d000) (1028K)]
> 167   0x00007f9328271300 JavaThread "Signal Dispatcher"          daemon [_thread_blocked, id=104815, stack(0x00007f92e760b000,0x00007f92e770c000) (1028K)]
> 168   0x00007f9328272b20 JavaThread "Service Thread"             daemon [_thread_blocked, id=104816, stack(0x00007f92e750a000,0x00007f92e760b000) (1028K)]
> 169   0x00007f9328274200 JavaThread "Monitor Deflation Thread"   daemon [_thread_blocked, id=104817, stack(0x00007f92e7409000,0x00007f92e750a000) (1028K)]
> 170   0x00007f9328275ef0 JavaThread "C2 CompilerThread0"         daemon [_thread_blocked, id=104818, stack(0x00007f92e7308000,0x00007f92e7409000) (1028K)]
> 171   0x00007f9328277780 JavaThread "C1 CompilerThread0"         daemon [_thread_blocked, id=104819, stack(0x00007f92e7207000,0x00007f92e7308000) (1028K)]
> 172   0x00007f932829f370 JavaThread "Notification Thread"        daemon [_thread_blocked, id=104820, stack(0x00007f92e7036000,0x00007f92e7137000) (1028K)]
> 173 Total: 9
> 174 
> 175 Other Threads:
> 176   0x00007f932822a3e0 VMThread "VM Thread"                           [id=104812, stack:(0x00007f92e7e7f000,0x00007f92e7f7f000) (1024K)]
> 177   0x00007f93282a1380 WatcherThread "VM Periodic Task Thread"        [id=104821, stack:(0x00007f92e2f00000,0x00007f92e3000000) (1024K)]
> 178   0x00007f932805add0 WorkerThread "GC Thread#0"                     [id=104807, stack:(0x00007f932c79c000,0x00007f932c89c000) (1024K)]
> 179   0x00007f932806c240 ConcurrentGCThread "G1 Main Marker"            [id=104808, stack:(0x00007f932c69b000,0x00007f932c79b000) (1024K)]
> 180   0x00007f932806d3b0 WorkerThread "G1 Conc#0"                       [id=104809, stack:(0x00007f932c59a000,0x00007f932c69a000) (1024K)]
> 181   0x00007f932812ea00 ConcurrentGCThread "G1 Refine#0"               [id=104810, stack:(0x00007f92fc1a6000,0x00007f92fc2a6000) (1024K)]
> 182   0x00007f932812fb60 ConcurrentGCThread "G1 Service"                [id=104811, stack:(0x00007f92fc0a5000,0x00007f92fc1a5000) (1024K)]
> 183 Total: 7
> ```
> 
> You sure you looked at the latest version?

Looks like we still have just one character difference:

`id=104819, stack(0x00007f92e7207000,0x00007f92e7308000)
`
vs

`id=104812, stack:(0x00007f92e7e7f000,0x00007f92e7f7f000)
`

tstuefe commented:
> If it were me I would also move the Java's thread running status at the end, especially because it is variable length, so that we would have:
> 
> ```
> 164 =>0x00007f9328028d10 JavaThread "main"                              [id=104806, stack(0x00007f932fdf0000,0x00007f932fef1000) (1028K), _thread_in_vm]
> ```
> 
> vs
> 
> ```
> 176   0x00007f932822a3e0 VMThread "VM Thread"                           [id=104812, stack:(0x00007f92e7e7f000,0x00007f92e7f7f000) (1024K)]
> ```
> 
> But I am not going to insist on that.

I'll remove the colon. But I don't want to change the java thread printout since there the chance is higher that someone is already parsing that.

gerard-ziemski reviewed (approved):
LGTM

dholmes-ora reviewed (approved):
Still good. Thanks

tstuefe commented:
/integrate
